### PR TITLE
Support MDS on ListDetectors page

### PR DESCRIPTION
**Pull request by jackiehanyang**
### Description

This pr adds MDS support on ListDetectors page. The default data source will be selected when MDS support is enabled.

Screenshot 2024-04-04 at 15 39 35

Screenshot 2024-04-08 at 12 41 56

### Issues Resolved

[List any issues this PR will resolve]

### Check List

- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
- [ ] Commits are signed per the DCO using `--signoff`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: public/pages/utils/helpers.ts
Comment (huyaboo) [inline public/pages/utils/helpers.ts:104]: If we are getting detectors to display, why are we mutating their `dataSourceId`s?

File: server/plugin.ts
Comment (huyaboo) [inline server/plugin.ts:90]: How will the `adService` behave in MDS world since `client` will be `undefined` here? I checked the `ad.ts` file and `client` is used as a `Local Cluster`.

**Review by huyaboo — COMMENTED** [review]
Could you please provide detailed steps with testing your changes? Does it still function with MDS disabled and/or hide local cluster enabled/disabled?

File: public/pages/utils/helpers.ts
Comment (jackiehanyang) [reply public/pages/utils/helpers.ts:104]: I'm not mutating their dataSourceId here, just setting the new dataSourceId field in Detector interface so they can be passed to the next page

File: public/pages/DetectorsList/containers/List/List.tsx
Comment (jackiehanyang) [reply public/pages/DetectorsList/containers/List/List.tsx:580]: This usage comes from the `Use Data Source Menu to Nav Bar or Data Source Selector inside Page` quip doc

File: public/models/interfaces.ts
Comment (jackiehanyang) [reply public/models/interfaces.ts:221]: dataSourceId will be `""` and fall back to use the existing client - https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/blob/8a94a9b027c2ad32b3a089a3899e04e25157577d/server/utils/helpers.ts#L105

File: server/plugin.ts
Comment (jackiehanyang) [reply server/plugin.ts:43]: will update this part

File: server/utils/helpers.ts
Comment (jackiehanyang) [reply server/utils/helpers.ts:89]: synced offline, tests will come later

File: server/plugin.ts
Comment (jackiehanyang) [reply server/plugin.ts:90]: in MDS world, the open search legacy client will be undefined and it will pickup the data source client as in MDS world dataSourceEnabled will always be true

File: server/plugin.ts
Comment (jackiehanyang) [reply server/plugin.ts]: updated in the latest commit

**Comment by BionIT** [comment]
Can we add some context and manual tests have been done for the PR?

**Review by BionIT — APPROVED** [review]

**Comment by huyaboo** [comment]
Approved but a couple of action items once this is merged:
1. Add some tests
2. Finalize with UX the behavior of AD Plugin if the `DataSourceSelectable` component returns to the callback function the invalid state (`[]`) and make that change in a subsequent PR.

**Review by huyaboo — APPROVED** [review]

**Comment by jackiehanyang** [comment]
> Can we add some context and manual tests have been done for the PR?

This pr just add the data source component on the list page and render the page according to data source selector. I will provide a record to include this page on the other pr I opened

**Comment by jackiehanyang** [comment]
> Approved but a couple of action items once this is merged:
> 
> 1. Add some tests
> 2. Finalize with UX the behavior of AD Plugin if the `DataSourceSelectable` component returns to the callback function the invalid state (`[]`) and make that change in a subsequent PR.

noted

File: server/utils/helpers.ts
Comment (ohltyler) [inline server/utils/helpers.ts:90]: nit: this can be typed as `RequestHandlerContext`

**Review by ohltyler — CHANGES_REQUESTED** [review]
Really great to understand the internals of this. Biggest suggestion here is to have a plan to refactor/support passing the selected data source ID through all of the node APIs. Details in comments

**Comment by ohltyler** [comment]
Also, let's make sure unit tests are passing. +1 to @huyaboo comment as well, maybe we can have a meta issue to make sure we have 1/ updated unit tests and 2/ a few integ tests for sanity testing. @huyaboo / @BionIT is there integ test framework support for mocking multiple data sources and invoking multiple cluster clients?

File: server/models/types.ts
Comment (jackiehanyang) [reply server/models/types.ts:92]: I'm adding the dataSourceId as port of this particular interface just for this list page. Because for list page, it uses a specific interface for its query params. Since dataSourceId also serves as a query param on this page, so I'm just adding it to this interface. For other pages, I'm appending the data source id the path itself. Please check the pr for detailed page - https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/pull/719/files#diff-cc594889ac44cb130233969adf3da57a18f360a2a77c57ee273aad5f39d6d733

Comment (ohltyler) [reply server/models/types.ts:92]: Understood, and cool, looks like some are adjusted that way there. Let's make all of them consistent through the node path the first time though. This will make it much cleaner and avoid making changes to this data model which doesn't logically fit.

File: public/pages/DetectorsList/containers/List/List.tsx
Comment (jackiehanyang) [reply public/pages/DetectorsList/containers/List/List.tsx:103]: addressed it in this commit - https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/pull/719/commits/6d8bd5bfcd265822d28dde119e22dd69ca12b7c5 in this pr - https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/pull/719/files

File: server/routes/ad.ts
Comment (jackiehanyang) [reply server/routes/ad.ts:637]: could you elaborate more on this? both `callWithRequest` and `getClientBasedOnDataSource` have no conflict with any existing imported fn names. I add the helper function `getClientBasedOnDataSource` in the helper file and calling it here.

File: server/utils/helpers.ts
Comment (jackiehanyang) [reply server/utils/helpers.ts:90]: addressed in this commit - https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/pull/719/commits/5057c0728b7ea9967312cfde0ad766f38e61dd85

**Comment by jackiehanyang** [comment]
> Also, let's make sure unit tests are passing. +1 to @huyaboo comment as well, maybe we can have a meta issue to make sure we have 1/ updated unit tests and 2/ a few integ tests for sanity testing. @huyaboo / @BionIT is there integ test framework support for mocking multiple data sources and invoking multiple cluster clients?

great suggestion. meta issue created - https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/issues/721

File: server/routes/ad.ts
Comment (ohltyler) [reply server/routes/ad.ts:637]: Whoops, yep you're right. No worries here, I confused it with `callAsCurrentUser()`

**Pull request closed without merging**
